src/makeStyles.tsx

```diff
@@ -48,44 +58,81 @@ export function createMakeStyles<Theme>(params: { useTheme: () => Theme }) {
             return function useStyles(params: Params) {
                 const theme = useTheme();
 
-                const { cx, css } = useCssAndCx();
+                const { css, cx } = useCssAndCx();
+
+                const cache = useTssEmotionCache();
 
                 return useMemo(() => {
+                    const refClassesCache: Record<string, string> = {};
+
+                    const refClasses = new Proxy<
+                        Record<
+                            RuleNameSubsetReferencableInNestedSelectors,
+                            string
+                        >
+                    >({} as any, {
+                        "get": (...args) => {
+                            const [, propertyKey] = args;
+
+                            if (typeof propertyKey === "symbol") {
+                                return Reflect.get(...args);
+                            }
+
+                            return (refClassesCache[propertyKey] = `${
+                                cache.key
+                            }-${outerCounter}${
+                                name !== undefined ? `-${name}` : ""
+                            }-${propertyKey}-ref`);
+                        },
+                    });
+
                     const cssObjectByRuleName = getCssObjectByRuleName(
                         theme,
                         params,
-                        (() => {
-                            let count = 0;
-
-                            return function createRef() {
-                                return `tss-ref${outerCounter}x${count++}${
-                                    label !== undefined ? `-${label}` : ""
-                                }`;
-                            };
-                        })(),
+                        refClasses,
                     );
 
                     const classes = objectFromEntries(
                         objectKeys(cssObjectByRuleName).map(ruleName => {
                             const cssObject = cssObjectByRuleName[ruleName];
 
-                            if (!cssObject.label) {
+                            if (!cssObject.name) {
                                 cssObject.label = `${
-                                    label !== undefined ? `${label}-` : ""
+                                    name !== undefined ? `${name}-` : ""
                                 }${ruleName}`;
                             }
 
-                            return [ruleName, css(cssObject)];
+                            return [
+                                ruleName,
+                                `${css(cssObject)}${
+                                    typeGuard<RuleNameSubsetReferencableInNestedSelectors>(
```

Comment:
You mean:

```typescript
ruleName in refClassesCache 
    ? ` ${refClassesCache[ruleName as RuleNameSubsetReferencableInNestedSelectors]}` 
    : ""
```

Yeah sure, it's functionally the same thing but using `typeGuard` is better  here.   
`ruleName` is of type `RuleNameSubsetReferencableInNestedSelectors` because `ruleName in refClassesCache` was true. If you use `as` you untie the unchecked assertion we made from the check that condition it.

Reply:
Well if the type guard were to do anything.
But in this case, it just uses the passed bool value as an assurance which is just the same as a cast

Reply:
[There is no arguing about that](https://www.typescriptlang.org/docs/handbook/advanced-types.html#type-guards-and-differentiating-types)